htmresearch/algorithms/synapse_learning.py

```diff
@@ -0,0 +1,241 @@
+                                     rng)
+
+
+  def learnOnNewSegments(self, connections, learningCells, growthCandidateAxons,
+                         rng, **kwargs):
+
+    numNewSynapses = min(len(growthCandidateAxons), self.maxSynapsesPerSegment)
+
+    newSegments = connections.createSegments(learningCells)
+    connections.growSynapsesToSample(newSegments, growthCandidateAxons,
+                                     numNewSynapses, self.initialPermanence,
+                                     rng)
+
+
+
+class ConnectToSampleOfActiveAxons(SynapseLearningRules):
+  """
+  Aim for "sampleSize" active synapses per segment. If a segment has fewer than
+  "sampleSize" active synapses, grow synapses to make up the difference.
+
+  Stay below a "maxSynapsesPerSegment". When a segment has
+  "maxSynapsesPerSegment" synapses, don't grow any new synapses on the segment.
+  """
+
+  def __init__(self, sampleSize, initialPermanence, permanenceIncrement,
+               permanenceDecrement, maxSynapsesPerSegment):
+    self.sampleSize = sampleSize
+    self.initialPermanence = initialPermanence
+    self.permanenceIncrement = permanenceIncrement
+    self.permanenceDecrement = permanenceDecrement
+    self.maxSynapsesPerSegment = maxSynapsesPerSegment
+
+
+  def learnOnExistingSegments(self, connections, learningSegments, activeAxons,
+                              growthCandidateAxons, rng, potentialOverlaps,
+                              **kwargs):
+
+    connections.adjustSynapses(learningSegments, activeAxons,
+                               self.permanenceIncrement,
+                               -self.permanenceDecrement)
+
+    maxNewBySegment = self.sampleSize - potentialOverlaps[learningSegments]
+
+    if self.maxSynapsesPerSegment != -1:
+      synapseCounts = connections.mapSegmentsToSynapseCounts(learningSegments)
+      numSynapsesToReachMax = self.maxSynapsesPerSegment - synapseCounts
+      maxNewBySegment = np.where(maxNewBySegment <= numSynapsesToReachMax,
+                                 maxNewBySegment, numSynapsesToReachMax)
+
+    connections.growSynapsesToSample(learningSegments, growthCandidateAxons,
+                                     maxNewBySegment, self.initialPermanence,
+                                     rng)
+
+
+  def learnOnNewSegments(self, connections, learningCells,
+                         growthCandidateAxons, rng, **kwargs):
+
+    numNewSynapses = min(len(growthCandidateAxons), self.sampleSize)
+
+    if self.maxSynapsesPerSegment != -1:
+      numNewSynapses = min(numNewSynapses, self.maxSynapsesPerSegment)
+
+    newSegments = connections.createSegments(learningCells)
+    connections.growSynapsesToSample(newSegments, growthCandidateAxons,
+                                     numNewSynapses, self.initialPermanence,
+                                     rng)
+
+
+def createSynapseLearningRules(**kwargs):
+  """
+  Factory for SynapseLearningRules
+  """
+  sampleSize = kwargs.get("sampleSize", -1)
```

Comment:
sampleSize and maxSynapsesPerSegment are determining the synapse rule to select. How this is done, and the meaning of those parameters, should be clearly documented.